providers/common/ciphers/aes.c

```diff
@@ -0,0 +1,424 @@
+        outlint = AES_BLOCK_SIZE;
+        out += AES_BLOCK_SIZE;
+    }
+    if (nextblocks > 0) {
+        if (!ctx->enc && ctx->pad && nextblocks == inl) {
+            if (!ossl_assert(inl >= AES_BLOCK_SIZE))
+                return 0;
+            nextblocks -= AES_BLOCK_SIZE;
+        }
+        outlint += nextblocks;
+        if (outsize < outlint)
+            return 0;
+        if (!ctx->ciph->cipher(ctx, out, in, nextblocks))
+            return 0;
+        in += nextblocks;
+        inl -= nextblocks;
+    }
+    if (!trailingdata(ctx->buf, &ctx->bufsz, AES_BLOCK_SIZE, &in, &inl))
+        return 0;
+
+    *outl = outlint;
+    return inl == 0;
+}
+
+static int aes_block_final(void *vctx, unsigned char *out, size_t *outl,
+                           size_t outsize)
+{
+    PROV_AES_KEY *ctx = (PROV_AES_KEY *)vctx;
+
+    if (ctx->enc) {
+        if (ctx->pad) {
+            padblock(ctx->buf, &ctx->bufsz, AES_BLOCK_SIZE);
+        } else if(ctx->bufsz == 0) {
+            *outl = 0;
+            return 1;
+        } else if(ctx->bufsz != AES_BLOCK_SIZE) {
+            /* TODO(3.0): What is the correct error code here? */
+            return 0;
+        }
+
+        if (outsize < AES_BLOCK_SIZE)
+            return 0;
+        if (!ctx->ciph->cipher(ctx, out, ctx->buf, AES_BLOCK_SIZE))
+            return 0;
+        ctx->bufsz = 0;
+        *outl = AES_BLOCK_SIZE;
+        return 1;
+    }
+
+    /* Decrypting */
+    /* TODO(3.0): What's the correct error here */
+    if (ctx->bufsz != AES_BLOCK_SIZE) {
+        if (ctx->bufsz == 0 && !ctx->pad) {
+            *outl = 0;
+            return 1;
+        }
+        return 0;
+    }
+
+    if (!ctx->ciph->cipher(ctx, ctx->buf, ctx->buf, AES_BLOCK_SIZE))
+        return 0;
+
+    /* TODO(3.0): What is the correct error here */
+    if (ctx->pad && !unpadblock(ctx->buf, &ctx->bufsz, AES_BLOCK_SIZE))
+        return 0;
+
+    if (outsize < ctx->bufsz)
+        return 0;
+    memcpy(out, ctx->buf, ctx->bufsz);
+    *outl = ctx->bufsz;
+    ctx->bufsz = 0;
+    return 1;
+}
+
+static int aes_stream_update(void *vctx, unsigned char *out, size_t *outl,
+                             size_t outsize, const unsigned char *in,
+                             size_t inl)
+{
+    PROV_AES_KEY *ctx = (PROV_AES_KEY *)vctx;
+
+    if (outsize < inl)
+        return 0;
+
+    if (!ctx->ciph->cipher(ctx, out, in, inl))
+        return 0;
+
+    *outl = inl;
+    return 1;
+}
+static int aes_stream_final(void *vctx, unsigned char *out, size_t *outl,
+                            size_t outsize)
+{
+    *outl = 0;
+    return 1;
+}
+
+static int aes_cipher(void *vctx, unsigned char *out, const unsigned char *in,
+                      size_t inl)
+{
+    PROV_AES_KEY *ctx = (PROV_AES_KEY *)vctx;
+
+    if (!ctx->ciph->cipher(ctx, out, in, inl))
+        return 0;
+
+    return 1;
+}
+
+#define IMPLEMENT_new_params(lcmode, UCMODE) \
+    static OSSL_OP_cipher_get_params_fn aes_##lcmode##_get_params; \
```

Comment:
This forward declaration isn't necessary -- the function is static and is on the following line.

Reply:
No, its strictly not necessary but I still think its a good idea (as suggested by @levitte). The point is it verifies that the function as declared is consistent with the `OSSL_OP_cipher_get_params_fn` typedef.

Reply:
I will push pretty hard to keep this kind of construct.  It does save us from stoopid mistakes, such as getting the arguments wrong.